HW2/P3/mandelbrot.pyx

```diff
@@ -27,22 +39,66 @@ cpdef mandelbrot(np.complex64_t [:, :] in_coords,
-                    if magnitude_squared(z) > 4:
-                        break
-                    z = z * z + c
-                out_counts[i, j] = iter
+    array_real = np.real(in_coords)
+    array_imag = np.imag(in_coords)
 
+    ones = AVX.float_to_float8(1)
+    twos = AVX.float_to_float8(2)
+    fours = AVX.float_to_float8(4)
 
+    with nogil, parallel(num_threads = 4):
+        for i in prange(in_coords.shape[0],schedule='static', chunksize=1):
+            thread_id = cython.parallel.threadid()
+            for j in range(0, in_coords.shape[1], 8):
+                c_real = AVX.make_float8(array_real[i, j + 7],
+                                    array_real[i, j + 6],
+                                    array_real[i, j + 5],
+                                    array_real[i, j + 4],
+                                    array_real[i, j + 3],
+                                    array_real[i, j + 2],
+                                    array_real[i, j + 1],
+                                    array_real[i, j + 0])
+
+                c_imag = AVX.make_float8(array_imag[i, j + 7],
+                                    array_imag[i, j + 6],
+                                    array_imag[i, j + 5],
+                                    array_imag[i, j + 4],
+                                    array_imag[i, j + 3],
+                                    array_imag[i, j + 2],
+                                    array_imag[i, j + 1],
+                                    array_imag[i, j + 0])
+                iter_now = AVX.make_float8(0, 0, 0, 0, 0, 0, 0, 0)
+                z_real = AVX.make_float8(0, 0, 0, 0, 0, 0, 0, 0)
+                z_imag = AVX.make_float8(0, 0, 0, 0, 0, 0, 0, 0)
+                for iter in range(max_iterations):
+                    # 1. calc mag
+                    temp_real = AVX.mul(z_real, z_real)
+                    temp_imag = AVX.mul(z_imag, z_imag)
+                    mag = AVX.add(temp_real, temp_imag)
+                    # 2. decide if continue
+                    b = AVX.greater_than(fours, mag)
+                    signs_int = AVX.signs(b)
+                    if signs_int == 0:
+                        break
+                    # 3. calculate new z: z = z * z + c
+                    temp_real = AVX.fmsub(z_imag, z_imag, c_real)
+                    temp_real = AVX.fmsub(z_real, z_real, temp_real)
+                    temp_imag = AVX.mul(z_real, z_imag)
+                    z_imag = AVX.fmadd(twos, temp_imag, c_imag)
+                    z_real = temp_real
+                    # 4. update iter
+                    iter_val = AVX.bitwise_and(b, ones)
+                    iter_now = AVX.add(iter_now, iter_val)
+                # 5. save to memory
+                AVX.to_mem(iter_now, &(temp[thread_id][0]))
+                # 6. type cast, cost a lot of time
+                for k in range(8):
+                    out_counts[i, j + k] = <int>(temp[thread_id][k])
```

Comment:
Comments make this code very readable (even though AVX is rather unintuitive/messy). Great job!